examples/simple_trainer_scaffold.py

```diff
@@ -0,0 +1,1117 @@
+    bilateral_grid_shape: Tuple[int, int, int] = (16, 16, 8)
+
+    # Enable depth loss. (experimental)
+    depth_loss: bool = False
+    # Weight for depth loss
+    depth_lambda: float = 1e-2
+
+    # Dump information to tensorboard every this steps
+    tb_every: int = 100
+    # Save training images to tensorboard
+    tb_save_image: bool = False
+
+    lpips_net: Literal["vgg", "alex"] = "alex"
+
+    def adjust_steps(self, factor: float):
+        self.eval_steps = [int(i * factor) for i in self.eval_steps]
+        self.save_steps = [int(i * factor) for i in self.save_steps]
+        self.max_steps = int(self.max_steps * factor)
+
+        strategy = self.strategy
+        strategy.refine_start_iter = int(strategy.refine_start_iter * factor)
+        strategy.refine_stop_iter = int(strategy.refine_stop_iter * factor)
+        strategy.voxel_size = self.voxel_size
+
+
+def create_splats_with_optimizers(
+    parser: Parser,
+    strategy: ScaffoldStrategy,
+    init_extent: float = 3.0,
+    init_opacity: float = 0.1,
+    init_scale: float = 1.0,
+    scene_scale: float = 1.0,
+    sh_degree: int = 3,
+    sparse_grad: bool = False,
+    batch_size: int = 1,
+    feature_dim: Optional[int] = None,
+    device: str = "cuda",
+    world_rank: int = 0,
+    world_size: int = 1,
+    voxel_size: float = 0.001,
+) -> Tuple[torch.nn.ParameterDict, Dict[str, torch.optim.Optimizer]]:
+
+    # Compare GS-Scaffold paper formula (4)
+    points = np.unique(np.round(parser.points / voxel_size), axis=0) * voxel_size
+    points = torch.from_numpy(points).float()
+
+    # Initialize the GS size to be the average dist of the 3 nearest neighbors
+    dist2_avg = (knn(points, 4)[:, 1:] ** 2).mean(dim=-1)  # [N,]
+    dist_avg = torch.sqrt(dist2_avg)
+    scales = torch.log(dist_avg * init_scale).unsqueeze(-1).repeat(1, 6)  # [N, 3]
+
+    # Distribute the GSs to different ranks (also works for single rank)
+    points = points[world_rank::world_size]
+    scales = scales[world_rank::world_size]
+    N = points.shape[0]
+    quats = torch.rand((N, 4))  # [N, 4]
+
+    features = torch.zeros((N, strategy.feat_dim))
+    offsets = torch.zeros((N, strategy.n_feat_offsets, 3))
+
+    opacities = torch.logit(torch.full((N, 1), init_opacity))  # [N,]
+    params = [
+        # name, value, lr
+        ("anchors", torch.nn.Parameter(points), 0.0001),
+        ("scales", torch.nn.Parameter(scales), 0.007),
+        ("quats", torch.nn.Parameter(quats), 0.002),
+        ("opacities_mlp", strategy.opacities_mlp.parameters(), 0.002),
+        ("features", torch.nn.Parameter(features), 0.0075),
+        ("opacities", torch.nn.Parameter(opacities), 5e-2),
+        ("offsets", torch.nn.Parameter(offsets), 0.01 * scene_scale),
+        ("colors_mlp", strategy.colors_mlp.parameters(), 0.008),
+        ("scale_rot_mlp", strategy.scale_rot_mlp.parameters(), 0.004),
+    ]
+
+    splats = torch.nn.ParameterDict({n: v for n, v, _ in params}).to(device)
```

Comment:
Having the MLPs to live inside the densification strategy class is a bit weird to me, as they (MLPs) are part of the representation of the scene, not the densification procedure. My proposal is to have a `ParameterDict` to host the anchor parameters and another `ModuleDict` to host the MLPs. Together they are the splats:
```
anchors: ParameterDict = ...
decoders: ModuleDict = ...
splats: {"anchors": anchors", "decoders": decoders}
```

Then, you could maintain two optimizers, one for the anchors and one for the MLPs. This way makes it easier for the densification strategy to update the parameters (you don't need to hard-code the key words in the `strategy/ops`):
```
optimizers = {"anchors": ..., "decoders": ...,}

strategy: ScaffoldStrategy = ...
strategy.step_pre_backward(params=splats["anchors"], optimizers=optimizers["anchors"])
...
```

For saving, I think you could just save splats as {"anchors": ..., "decoders": ....}.

Reply:
I addressed your comments (hopefully). Please check.